src/probability/bell_inequality.lean

```diff
@@ -0,0 +1,96 @@
+
+Consider two observers, $$A$$ and $$B$$, performing measurements that take values in $${-1, 1}$$.
+
+Let `ℙ` be a probability measure, and let `A i` and `B i` for $$i ∈ {0, 1, 2}$$ be random
+variables taking values in $${-1, 1}$$. For convenience, we identify $${-1, 1}$$ with `ℤˣ`. `A i`
+represent the outcomes of measurements done by $$A$$, and `B i` those done by $$B$$. We assume
+perfect anticorrelation between the outcomes of $$A$$ and $$B$$: 𝔼[(A i) (B i)] = -1. Bell's
+inequality states that `𝔼[(A 1) (B 2)] - 𝔼[(A 1) (B 2)] ≤ 1 + 𝔼[(A 2) (B 0)]`.
+
+## TODO
+
+J.S. Bell generalized the inequality in 1975 to include more probabilistic theories. In particular,
+the 1975 statement does not require observables to simultaneously have values. Instead, it is solely
+based on the requirement of locality. The 1975 inequality thus gives a more direct proof of Bell's
+theorem.
+
+## References
+
+* [J.S. Bell, *On the Einstein Podolsky Rosen Paradox*, 1964][MR3790629]
+* [J.S. Bell, *The theory of local beables*, 1975,
+  reproduced in chapter 7 of *Speakable and unspeakable in quantum mechanics*][MR915338]
+
+## See also
+
+`CHSH_inequality_of_comm` is a star-algebra version of Bell's inequality.
+-/
+
+open filter measure_theory
+
+variables {Ω : Type*} [measurable_space Ω] {ℙ : measure Ω} [is_probability_measure ℙ] {f g : Ω → ℤˣ}
+  {A B : fin 3 → Ω → ℤˣ}
+
+private lemma norm_aux (a : ℤˣ) : ‖(a : ℝ)‖ ≤ 1 :=
+by obtain rfl | rfl := int.units_eq_one_or a; simp
+
+/-- The precise version of the CHSH inequality we need. -/
+private lemma CHSH_aux (A₁ A₂ B₀ B₂ : ℤˣ) :
+  (A₁ : ℝ) * B₂ - A₁ * B₀ - A₂ * B₂ ≤ 1 + A₂ * B₀ + 1 :=
+by obtain rfl | rfl := int.units_eq_one_or A₁; obtain rfl | rfl := int.units_eq_one_or A₂;
+  obtain rfl | rfl := int.units_eq_one_or B₀; obtain rfl | rfl := int.units_eq_one_or B₂; norm_num
+
+private lemma ae_strongly_measurable_aux (hf : measurable f) :
+  ae_strongly_measurable (λ ω, (f ω : ℝ)) ℙ :=
+begin
+  refine (measurable.comp (λ s hs, _) hf).ae_strongly_measurable,
+  exact ⟨coe ⁻¹' s, trivial, rfl⟩,
+end
+
+private lemma integrable_aux (hf : measurable f) : integrable (λ ω, (f ω : ℝ)) ℙ :=
+⟨ae_strongly_measurable_aux hf, has_finite_integral_of_bounded $ eventually_of_forall $ λ _,
+  norm_aux _⟩
+
+private lemma integrable_mul_aux (hf : measurable f) (hg : measurable g) :
+  integrable (λ ω, (f ω * g ω : ℝ)) ℙ :=
+(integrable_aux hg).bdd_mul (ae_strongly_measurable_aux hf) ⟨1, λ _, norm_aux _⟩
+
+/-- **Bell's inequality (1964 version)** Given six random variables `A B : fin 3 → Ω → ℤˣ` taking
+values in `±1`, and assuming perfect anticorrelation on the diagonal (that is, `𝔼[(A i) (B i)] = -1`
+for all `i`), we have that `𝔼[(A 1) (B 2)] - 𝔼[(A 1) (B 0)] ≤ 1 + 𝔼[(A 2) (B 0)]`. -/
+theorem bell_inequality_1964 (ha : ∀ i, measurable (A i)) (hb : ∀ i, measurable (B i))
+  (anticorrelation : (∫ ω, A 2 ω * B 2 ω ∂ℙ : ℝ) = -1) :
+  (∫ ω, A 1 ω * B 2 ω ∂ℙ : ℝ) - ∫ ω, A 1 ω * B 0 ω ∂ℙ ≤ 1 + ∫ ω, A 2 ω * B 0 ω ∂ℙ :=
```

Comment:
Sorry, I am really confused here. Why does one even talk about the `B i`? Isn't `B i = - A i a.e.`? Surely this is clearer with just the three variables. Indeed this seems to be what Bell wrote (eq 15 of https://journals.aps.org/ppf/pdf/10.1103/PhysicsPhysiqueFizika.1.195).

(Except for the missing absolute value, where did that go?)

Reply:
Yes, `B i = - A i` almost everywhere, but the experimental setup considered by Bell is to have a measurement done by A and one by B, so the expectation of Ai*Bj is really what we are after. In other words, A and B perform their measurements, and we look at the statistics of the product of their outcomes. In Bell's paper, the inequality is (15), with P defined in (2) (which is shown to be the same as taking Bi=-Ai in (14).

As for the absolute value, the other inequality is obtained from renaming A'_1=-A_1 and applying the same inequality. I can add that statement to the theorem today or tomorrow.

Reply:
I added the absolute value statement. As for `B i` vs `-A i`, we never have in mathlib a statement with two ae equal functions not substituted for each other, because we have congruency lemmas for those cases.